components/datetime/src/format/time_zone.rs

```diff
@@ -37,12 +37,77 @@ where
     T: TimeZoneInput,
 {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write_pattern(self.time_zone_format, self.time_zone, f).map_err(|_| core::fmt::Error)
+        write_zone(self.time_zone_format, self.time_zone, f).map_err(|_| core::fmt::Error)
     }
 }
 
-// TODO(#622) Make this public once TimeZoneFormat is public.
-pub(crate) fn write_pattern<T, W>(
+pub(crate) fn write_zone<T, W>(
+    time_zone_format: &TimeZoneFormat,
+    time_zone: &T,
+    w: &mut W,
+) -> Result<(), Error>
+where
+    T: TimeZoneInput,
+    W: fmt::Write + ?Sized,
+{
+    match (&time_zone_format.patterns, &time_zone_format.config) {
+        (Some(_), None) => write_pattern(time_zone_format, time_zone, w),
+        (None, Some(_)) => write_config(time_zone_format, time_zone, w),
+        _ => todo!("Invalid state (both populated, or none populated, shouldn't be possible)"),
```

Comment:
> issue

Whoops, I meant to flag this on my previous pass, but it looks like I accidentally skipped it.

This should be `unreachable!()` rather than `todo!()`. 

And we should also be more specific here about what was populated. 

In fact, we should probably be specific about both cases:

```rust
(Some(_), Some(_)) => unreachable!("TimeZoneFormat should have a configuration or a pattern, but found both"),
(None, None) => unreachable!("TimeZoneFormat should have either a configuration or a pattern, but found neither"),
```